src/ext/detective.lisp

```diff
@@ -0,0 +1,282 @@
+
+   :search-references
+
+   :capture-references
+
+   :navigate-reference
+
+   :move-to-reference)
+  
+  #+sbcl
+  (:lock t))
+
+(in-package :lem/detective)
+
+(defclass reference ()
+  ((name :type string 
+         :initarg :reference-name
+         :reader reference-name)
+
+   (point :initarg :reference-point
+          :reader reference-point)))
+
+(defclass function-reference (reference)
+  ((arguments :type list
+              :initarg :function-arguments
+              :reader function-arguments)))
+
+(defclass package-reference (reference)
+  ((use :type list
+        :initarg :package-use
+        :reader package-use )
+   (export :type list
+           :initarg :package-export
+           :reader package-export)))
+
+(defclass class-reference (reference)
+  ((parents :type list
+            :initarg :class-parents
+            :reader class-parents)
+   (attributes :type list
+               :initarg :class-attributes
+               :reader class-attributes)))
+
+(defclass variable-reference (reference)
+  ((value :initarg :variable-reference-value
+          :reader variable-reference-value)))
+
+
+(defclass misc-reference (reference)
+  ((custom-type :initarg :misc-custom-type
+                :type string
+                :reader misc-custom-type)))
+
+(defstruct capture-regex regex function)
+
+(defclass detective-search () ())
+
+(defclass search-regex (detective-search)
+  ((function-regex :type (or capture-regex null)
+                   :initform nil
+                   :initarg :function-regex
+                   :writer set-function-regex
+                   :reader search-function-regex)
+   (package-regex :type (or capture-regex null)
+                  :initform nil
+                  :initarg :package-regex
+                  :writer set-package-regex
+                  :reader search-package-regex)
+   (class-regex :type (or capture-regex null)
+                :initform nil
+                :initarg :class-regex
+                :writer set-class-regex
+                :reader search-class-regex)
+   (variable-regex :type (or capture-regex null)
+                   :initform nil
+                   :initarg :variable-regex
+                   :writer set-variable-regex
+                   :reader search-variable-regex )
+   (misc-regex :type (or capture-regex null)
+               :initform nil
+               :initarg :misc-regex
+               :writer set-misc-regex
+               :reader search-misc-regex)))
+
+
+(defun buffer-references (buffer)
+  (buffer-value buffer 'references))
+
+(defun (setf buffer-references) (value buffer)
+  (setf (buffer-value buffer 'references) value))
+
+(defgeneric search-references (search-class))
+
+(defmethod search-references ((search search-regex))
+  (labels ((find-ref (regex class)
+             (with-point ((p (buffer-start-point (point-buffer (current-point)))))
+               (loop :for position = (search-forward-regexp p (capture-regex-regex regex))
```

Comment:
In lem, the name position is often used for an integer (e.g. `cl:position` function), and point is often used to represent a `lem:point`.
I think it is less confusing to use the variable point here.

Reply:
Done :+1: